src/axom/core/ArrayBase.hpp

```diff
@@ -423,8 +471,35 @@ class ArrayBase
   /*! \brief Test if idx is within bounds */
   AXOM_HOST_DEVICE inline bool inBounds(IndexType idx) const
   {
-    return idx >= 0 && idx < asDerived().size();
+    return idx >= 0 && idx < memorySize();
   }
+
+  /// \brief Checks a shape and stride array for correct bounds.
+  AXOM_HOST_DEVICE inline void validateShapeAndStride(
+    const StackArray<IndexType, DIM>& shape,
+    const StackArray<IndexType, DIM>& stride)
+  {
+    int sorted_dims[DIM];
+    for(int dim = 0; dim < DIM; dim++)
+    {
+      sorted_dims[dim] = dim;
+    }
+    // Sort the dimensions by stride.
+    axom::utilities::insertion_sort(sorted_dims,
+                                    DIM,
+                                    [&](int dim_a, int dim_b) -> bool {
+                                      return stride[dim_a] < stride[dim_b];
+                                    });
+    // Work from the smallest-strided dimension to the largest-strided.
+    for(int dim = 0; dim < DIM - 1; dim++)
+    {
+      int minor_dim = sorted_dims[dim];
+      int major_dim = sorted_dims[dim + 1];
+      assert(stride[major_dim] >= stride[minor_dim] * shape[minor_dim]);
+      assert(stride[major_dim] % stride[minor_dim] == 0);
```

Comment:
Should we use SLIC_ERROR_IF?  This error can be caused by the user.

Reply:
We can't use `slic` macros in `core` since `slic` depends on `core`.

See other comments in this PR for more context.